RecoCTPPS/TotemRPLocal/interface/FastLineRecognition.h

```diff
@@ -0,0 +1,121 @@
+/****************************************************************************
+*
+* This is a part of TOTEM offline software.
+* Authors: 
+*   Jan Kašpar (jan.kaspar@gmail.com) 
+*
+****************************************************************************/
+
+#ifndef RecoCTPPS_TotemRPLocal_FastLineRecognition
+#define RecoCTPPS_TotemRPLocal_FastLineRecognition
+
+#include "DataFormats/Common/interface/DetSet.h"
+#include "DataFormats/Common/interface/DetSetVector.h"
+
+#include "Geometry/VeryForwardGeometryBuilder/interface/TotemRPGeometry.h"
+#include "DataFormats/CTPPSReco/interface/TotemRPRecHit.h"
+#include "DataFormats/CTPPSReco/interface/TotemRPUVPattern.h"
+
+//#define TUNE 1
+
+#ifdef TUNE
+  #include "TH1D.h"
+#endif
+
+
+/**
+ * \brief Class performing optimized hough transform to recognize lines.
+**/
+
+class FastLineRecognition
+{
+  protected:
+    /// the uncertainty of 1-hit cluster, in mm
+    static const double sigma0;
+
+    /// "typical" z
+    double z0;
+
+    /// cluster half widths in a and b
+    double chw_a, chw_b;
+
+    /// weight threshold for accepting pattern candidates (clusters)
+    double threshold;
+
+    /// pointer to the geometry
+    const TotemRPGeometry* geometry;
+
+    struct GeomData
+    {
+      double z;   ///< z position of a sensor (wrt. IP)
+      double s;   ///< sensor's centre projected to its read-out direction
+    };
+
+    /// map: raw detector id --> GeomData
+    std::map<unsigned int, GeomData> geometryMap;
+
+    /// expects raw detector id
+    GeomData GetGeomData(unsigned int id);
+
+    struct Point
+    {
+      unsigned int detId;       ///< raw detector id
+      const TotemRPRecHit* hit; ///< pointer to original reco hit
+      double h;                 ///< hit position in global coordinate system
+      double z;                 ///< z position with respect to z0
+      double w;                 ///< weight
+      bool usable;              ///< whether the point can still be used
+      Point(unsigned int _d=0, const TotemRPRecHit* _hit=NULL, double _h=0., double _z=0., double _w=0.) :
+        detId(_d), hit(_hit), h(_h), z(_z), w(_w), usable(true) {}
+    };
+    
+    /// cluster of intersection points
+    struct Cluster
+    {
+      double Saw, Sbw, Sw, S1;
+      double weight;
+      double min_a, max_a, min_b, max_b;
+
+      std::vector<const Point *> contents;
+      
+      Cluster() : Saw(0.), Sbw(0.), Sw(0.), S1(0.), weight(0.)
+#ifdef TUNE
+                  , min_a(1E100), max_a(-1E100),
+                  min_b(1E100), max_b(-1E100)
+#endif
+                      {}
+      
+      void Add(const Point *p1, const Point *p2, double a, double b, double w);
```

Comment:
methods and data members should start with lower case (we noticed this few weeks ago)